File: cpp/02_datatypes.cpp
// ============================================================
// Topic:  Data Types
// Summary: Core scalar types, nullptr, typeid, casting, stringstream.
// Run:     g++ -std=c++17 cpp/02_datatypes.cpp -o cpp/02_datatypes && cpp/02_datatypes
// ============================================================

#include <iostream>
#include <string>
#include <sstream>
#include <typeinfo>

int main() {
    // ----- 1. The core scalar types -----
    int an_int = 42;
    double a_float = 3.14;
    std::string a_str = "hello";
    bool a_bool = true;
    char a_char = 'A';
    int* nothing = nullptr;  // C++ null pointer   (Python: None, TS: null)
    std::cout << an_int << " " << a_float << " " << a_str << " "
              << std::boolalpha << a_bool << " " << a_char << " "
              << (nothing == nullptr ? "nullptr" : "ptr") << std::endl;
    // Output: 42 3.14 hello true A nullptr

    // ----- 2. Checking a type (typeid) -----
    std::cout << typeid(an_int).name() << std::endl;
    // Output: i  (implementation-defined; often mangled like "int" on MSVC)
    std::cout << typeid(a_str).name() << std::endl;
    // Output: NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE  (or similar)

    // ----- 3. static_cast — explicit conversion (no implicit coercion) -----
    double pi = 3.99;
    int truncated = static_cast<int>(pi);
    std::cout << truncated << std::endl;
    // Output: 3

    std::string num_text = "10";
    int parsed = std::stoi(num_text);
    std::cout << parsed + 5 << std::endl;
    // Output: 15

    std::cout << std::to_string(10) + "5" << std::endl;
    // Output: 105

    // ----- 4. stringstream — build/parse strings -----
    std::ostringstream oss;
    oss << "value=" << 42;
    std::cout << oss.str() << std::endl;
    // Output: value=42

    std::istringstream iss("3.5");
    double from_stream;
    iss >> from_stream;
    std::cout << from_stream << std::endl;
    // Output: 3.5

    // ----- 5. Numeric detail -----
    std::cout << 7 / 2 << std::endl;       // integer division   Output: 3
    std::cout << 7.0 / 2 << std::endl;     // float division      Output: 3.5
    std::cout << 10 % 3 << std::endl;      // modulo              Output: 1

    // ----- 6. bool converts only from bool expressions -----
    std::cout << std::boolalpha << (an_int != 0) << " "
              << (!a_str.empty()) << std::endl;
    // Output: true true
    // Note: unlike Python, non-zero int is NOT implicitly truthy in if() — compare explicitly.

    // Cross-language note:
    //   C++:  int / double / bool / char / std::string / nullptr
    //   Python: int / float / bool / str / None  (dynamic; type() at runtime)
    //   TS:   number / boolean / string / null / undefined
    //   Rust: i32 / f64 / bool / char / String / None (Option)
    //   C#:   int / double / bool / char / string / null
    //   C++ is statically typed; use static_cast or stoi/stod for conversions, not silent coercion.
    return 0;
}

File: typescript/02_datatypes.ts
// ============================================================
// Topic:  Data Types
// Summary: primitives, typeof, null vs undefined, casting.
// Run:     npx ts-node typescript/02_datatypes.ts
// ============================================================

// ----- 1. Primitive types (note: ONE number type, no int/float split) -----
let n: number = 42; // also 3.14 — number covers both (Python: int AND float)
let str: string = "hello"; // Python: str
let flag: boolean = true; // Python: True/False (lowercase here!)
let nothing: null = null; // Python: None
let missing: undefined = undefined; // Python has NO equivalent (only None)
console.log(n, str, flag, nothing, missing);
// Output: 42 hello true null undefined

// ----- 2. Checking a type -> typeof  (Python: type(x)) -----
console.log(typeof n); // Output: number
console.log(typeof str); // Output: string
console.log(typeof flag); // Output: boolean
console.log(typeof undefined); // Output: undefined

// ----- 3. Collection types -----
const arr: number[] = [1, 2, 3]; // array        (Python: list)
const obj = { k: "v" }; // object       (Python: dict)
const map = new Map([["k", "v"]]); // Map          (Python: dict)
const set = new Set([1, 2, 3]); // Set          (Python: set)
console.log(arr, obj, map, set);

// ----- 4. Conversion / casting -----
console.log(Number("10") + 5); // 15    | Python: int("10") + 5
console.log(String(10) + "5"); // "105" | Python: str(10) + "5"
console.log(parseInt("3.9")); // 3     (parse to int)
console.log(parseFloat("3.5")); // 3.5
console.log(Boolean(0), Boolean(""), Boolean(1)); // false false true

// ----- 5. The two "empty" values (Python only has None) -----
let a; // undefined: declared but never assigned
let b = null; // null: intentionally empty
console.log(a, b); // Output: undefined null

// ----- 6. Special TS types -----
let anything: any = 5; // disables checking (avoid)
anything = "now a string"; // allowed with any
let unknownVal: unknown = 10; // safer "any"; must narrow before use
console.log(anything, unknownVal); // Output: now a string 10

// Cross-language note:
//   TS has ONE `number` (a 64-bit float). Python splits int vs float.
//   TS has TWO empties: null + undefined. Python has just None.
